### ISSUE TITLE
About bpf_xdp_adjust_tail

### ISSUE THREAD

**Issue by glennWang** [xdp-project/xdp-tutorial#234]
I'm not sure my understanding is right about `bpf_xdp_adjust_tail`.

`bpf_xdp_adjust_tail` is able to enlarge the underlying packet buffer.
For example, receive a 100 byte size ​udp packet on xdp, I can enlarge the packet buffer as 1000 byte and assemble the reply packet on xdp, and then return XDP_TX.

When enlarging the packet, if there be some limit ?
For example,  there is a max limit, only can grow 256 byte or 1000 byte.

**Comment by tohojo** [comment]
Yes, there's a max, but it will be driver-dependent exactly what that
limit is; and you'll be limited by the MTU of course...

**Comment by glennWang** [comment]
@tohojo @netoptimizer 

a strange problem：

xdp receive udp and reply.

ctx->data_end - ctx->data == 92

and call `bpf_xdp_adjust_tail(ctx , grow_val)`

when grow_val <= 356, grow ok. But, when grow_val  >= 357, it would fail.

kernel version: 
`Linux fedora33 5.13.9-150.vanilla.1.fc33.x86_64 #1 SMP Sun Aug 8 11:37:31 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
`

**Comment by tohojo** [comment]
In skb mode the size is limited by whatever is available in the skb. Which I guess is 356 bytes in your case :)

**Comment by glennWang** [comment]
> In skb mode the size is limited by whatever is available in the skb. Which I guess is 356 bytes in your case :)


I'm confused. 

When I have a try on physical machine( ubuntu20.04 with kernel version 5.8.0)，Native mode and SKB mode all is ok calling `bpf_xdp_adjust_tail(ctx , grow_val)`, grow_val > 357. 

But, when I try again on my virtual machine with SKB mode (virtualbox ubuntu20.04 with kernel 5.8.0 and fedora31 with kernel 5.8.18). when grow_val > 357, it would fail.

Is there any params can be config to solve the problem?

**Comment by tohojo** [comment]
Not too surprising that different drivers and setups results in different allocation sizes. I don't think there's any way to influence that either. You may be able to use TC BPF instead to resize the skb though...

**Comment by glennWang** [comment]
> Not too surprising that different drivers and setups results in different allocation sizes. I don't think there's any way to influence that either. You may be able to use TC BPF instead to resize the skb though...

But, xdp on SKB mode is called after skb allocation, it should not be limit by different drivers in principle .

**Comment by tohojo** [comment]
skbs are allocated in the drivers, though. The allocation size can depend on the packet size as well, so I suspect you'd get different results if you test with larger packets :)

**Comment by glennWang** [comment]
> skbs are allocated in the drivers, though. The allocation size can depend on the packet size as well, so I suspect you'd get different results if you test with larger packets :)

In my scene, want to implement a dns-cache on xdp. xdp receive a query, and then lookup bpf map and reply directly after call bpf_xdp_adjust_tail().  The packet is not very big.  maybe, bpf_xdp_adjust_tail(ctx, 512-query_pkt_size) is the biggest.
Or, I make some error when using bpf_xdp_adjust_tail.

**Comment by tohojo** [comment]
Right, but why are you using XDP in skb mode? Max frame size should also be more stable in native mode...

**Comment by glennWang** [comment]
> Right, but why are you using XDP in skb mode? Max frame size should also be more stable in native mode...

Service will be deployed on cloud VM. SKB mode is more friendly for VM.

**Comment by tohojo** [comment]
glennWang ***@***.***> writes:

> Service will be deployed on cloud VM. SKB mode is more friendly for
> VM.

But also doesn't realise the performance gains of XDP - you might as
well implement it as TC-BPF, then, and use bpf_skb_adjust_room() to
properly resize skbs...

**Comment by akiyano** [comment]
@tohojo,
As you said the extent to which we can increase the size of a packet is driver dependant.
More specifically, it depends on how much tailroom is left in the skb (for generic mode) or in the xdp_buff (for native mode.)
There  is a #define XDP_PACKET_HEADROOM =256 in the kernel, to make sure there is enough headroom in xdp packets.
Wouldn't it be a good idea to add a  XDP_PACKET_TAILROOM= <some number>?
Or maybe add a driver hook (ethtool/devlink) that sets the extra tailroom, which the driver needs to reserve.
If such a standard is defined, increasing the packet size using bpf_xdp_adjust_tail will work on as many platforms as possible.

**Comment by tohojo** [comment]
Well, such a define would end up being the lowest common denominator between all driver implementations so I'm not sure it's terribly useful at this point? It *does* work on as many platforms as possible: it's not like there are any artificial limits put in place, it's just that some drivers/platforms have optimisations in place that limits the available space, which is sort of a fundamental constraint...

**Comment by akiyano** [comment]
Thanks for your quick answer.
I totally agree, which is why I made my second suggestion.

It only makes sense for a driver to use as little space as possible.
Therefore many drivers will not have much tailroom for bpf_xdp_adjust_tail() to be usefull for increasing packet sizes.

To make bpf_xdp_adjust_tail() more usefull I suggest adding such extra tailroom when needed, by adding a standard knob that can be controlled by the user (via ethtool/devlink).
This way if a user wants to use bpf_xdp_adjust_tail() to increase buffer size, he will be able to make sure the driver leaves enough space for it to work. On all other cases the driver will not do it, saving some space.

**Comment by tohojo** [comment]
What would the driver do with such a request, though? Say the user
requests 500 bytes of tailroom, but the memory allocation model used by
the driver means there's only MTU+250 bytes of space left in each page.
What should the driver do in this case?

**Comment by akiyano** [comment]
This is technically solvable depending on how important the bpf_xdp_adjust_tail() use case is.

For example simplest solution: 

Lets say the driver gives 4K buffers to the NIC for RX packets.
In case we want to have 500 bytes of tailroom we instead give 4K-500 buffer size to the NIC.

In this solution we lose 500 bytes per buffer which is a drawback, but if the use case of bpf_xdp_adjust_tail() increasing the packet is more important than this drawback, then it will be used.

Next step of such a solution can be adding a feature to the NIC HW, so it is the NIC resposibility to make sure there is at least 500 bytes left in the last buffer. With such a NIC feature, we dont lose 500 bytes for all non-last buffers.

**Comment by tohojo** [comment]
But we're not allocating arbitrary sized buffers, fox XDP we're
allocating full pages; there is no additional limit on the size of the
data that XDP can put in there, the limit on adjust_tail is just the
size of the data page.

The only exception to this is that there are some Intel drivers which
split each page and stuffs two packets into each page, limiting the
packet size to 2k-overhead instead of 4k-overhead. This was a deliberate
decision on the driver developers to conserve system memory, but again,
XDP does not impose any buffer size limits outside of this. So a user
configuration knob wouldn't really do anything in either of these cases...

**Comment by yifan513** [comment]
Thank your guys' discussion. I met the exact same issue with bpf_xdp_adjust_tail.
So any good example for TC-BPF?
Thanks.
